### DISCUSSION TITLE
Run server in a background thread (not as scheduled task)

### DISCUSSION THREAD

**Discussion by banesullivan** [Kitware/trame#173] [Q&A]
Would it be possible to run the trame server in a background thread? My attempt looks something like:

```py
from trame.app import get_server
import threading


class ServerThread(threading.Thread):
    def __init__(
        self,
        server,
        start=True,
    ):
        self._server = server
        threading.Thread.__init__(self, daemon=True, target=self.server_start)

        if start:
            self.start()

    def server_start(self):
        self._server.start(
            exec_mode="main",
            port=0,
            open_browser=False,
            show_connection_info=False,
            disable_logging=True,
            timeout=0,
            # thread=True,
        )
        
server = get_server('foo')
thread = ServerThread(server)
thread
```

However, `server.start(exec_mode="main")` is non-blocking and the thread immediately finalizes.

Any ideas on how to do this sort of thing?

My goal is not to schedule a coroutine that starts the server or have a task that runs when available. (for example, this is needed for single-cell execution in a Jupyter notebook)

**Comment by jourdain** [comment]
I think this is because the event loop is running in a thread, you need to tight it up with it as technically the `exec_mode="main"` is blocking... 

Something [like that](https://stackoverflow.com/questions/52298922/how-do-i-set-the-asyncio-event-loop-for-a-thread-in-python) needs to happen.

**Reply by banesullivan** [reply]
Well, now I can launch it and keep the thread running, but the `running_stage` is hung on 1:

```py
from trame.app import get_server
import threading
import asyncio


class ServerThread(threading.Thread):
    def __init__(
        self,
        server,
        start=True,
    ):
        self._event_loop = asyncio.new_event_loop()
        self._server = server
        threading.Thread.__init__(self, daemon=True, target=self.launch)

        if start:
            self.start()

    def server_start(self):
        self._server.start(
            exec_mode="main",
            port=0,
            open_browser=False,
            show_connection_info=False,
            disable_logging=True,
            timeout=0,
        )
        
    def launch(self):
        asyncio.set_event_loop(self._event_loop)
        asyncio.get_event_loop().call_soon(self.server_start)
        self._event_loop.run_forever()
        
server = get_server('foo')
thread = ServerThread(server)
```

```py
>>> server._running_stage
1
```

**Reply by jourdain** [reply]
I don't think you can use `exec_mode="main",` if you call yourself `self._event_loop.run_forever()`.
Did you try to run the trame server as a `task`? [All options](https://github.com/Kitware/wslink/blob/master/python/src/wslink/server.py#L292-L294)

**Reply by banesullivan** [reply]
Hm. I feel like I'm close with this:

```py
from trame.app import get_server
import threading
import asyncio


class ServerThread(threading.Thread):
    def __init__(
        self,
        server,
    ):
        if server._running_stage != 0:
            raise RuntimeError('Server already in a running stage')
        
        self._event_loop = asyncio.new_event_loop()
        self._server = server
        threading.Thread.__init__(self, daemon=True, target=self.launch)
        
        def finalize(future):
            print('finalized', self._server._running_stage)
            self._event_loop.close()

        self.start()
        future =  asyncio.run_coroutine_threadsafe(self.server_coroutine(), self._event_loop)
        future.add_done_callback(finalize)

    def server_coroutine(self):
        main_loop = asyncio.get_event_loop()
        asyncio.set_event_loop(self._event_loop)
        coroutine = self._server.start(
            exec_mode="coroutine",
            port=0,
            open_browser=False,
            show_connection_info=False,
            disable_logging=True,
            timeout=0,
            thread=True,
        )
        asyncio.set_event_loop(main_loop)
        return coroutine

    def launch(self):
        asyncio.set_event_loop(self._event_loop)
        self._event_loop.run_forever()
        
server = get_server('foo')
thread = ServerThread(server)
```

The server launches within the thread and makes it to running stage 2 but as soon as I try to use that server with VTK, it kills the kernel 😕

**Reply by banesullivan** [reply]
This works for client-side rendering but not server-side rendering with VTK.

**Reply by jourdain** [reply]
So for doing rendering on a different thread, they are some complication but we have some path to support that in VTK since we are running into some of those limitation with the ParaView Async project.

@jspanchu might be able to tell you more. But on mac, he had to fix something in VTK which got reverted due to some side effect.

**Reply by jspanchu** [reply]
Yup, all Cocoa (the windowing system library on macOS) API must be invoked in the main thread. In your case it's the thread that entered [`int main()`:python.c](https://github.com/python/cpython/blob/main/Programs/python.c#L13). Such GUI calls include creating and updating a render window's OpenGL context. I'm almost sure the kernel is killed during creation of the render window. @banesullivan Do you have a stack trace?

The VTK fix got reverted because it broke tiled render windows.

The ParaView Async project gets around by queuing up all Cocoa commands on the server main thread and executing them there. It's difficult to bring it to VTK because the workaround heavily depends on the `rxcpp` library that we use for reactive programming in ParaView Async.
